src/coprocessor/endpoint.rs

```diff
@@ -60,194 +65,174 @@ const OUTDATED_ERROR_MSG: &str = "request outdated.";
+                    Ok(ctx) => ctx,
+                    Err(e) => return on_resp.respond(err_resp(e)),
+                };
+                if !on_resp.is_streaming() {
+                    let batch_row_limit = self.batch_row_limit;
+                    let do_request = move || {
+                        tracker.record_wait();
+                        let mut resp = ctx.handle_request(batch_row_limit).unwrap_or_else(err_resp);
+                        let mut cop_stats = CopStats::default();
+                        ctx.collect_statistics_into(&mut cop_stats.stats);
+                        ctx.collect_metrics_into(&mut cop_stats.scan_counter);
+                        tracker.record_handle(&mut resp, cop_stats);
+                        future::ok::<_, ()>(on_resp.respond(resp))
+                    };
+                    return pool.spawn_fn(do_request).forget();
+                }
+                // For streaming.
+                let batch_row_limit = self.stream_batch_row_limit;
+                let s = stream::unfold((ctx, false), move |(mut ctx, finished)| {
+                    if finished {
+                        return None;
+                    }
+                    tracker.record_wait();
+                    let (item, finished) = ctx.handle_streaming_request(batch_row_limit)
+                        .unwrap_or_else(|e| (Some(err_resp(e)), true));
+                    item.map(|mut resp| {
+                        let mut cop_stats = CopStats::default();
+                        ctx.collect_statistics_into(&mut cop_stats.stats);
+                        ctx.collect_metrics_into(&mut cop_stats.scan_counter);
+                        tracker.record_handle(&mut resp, cop_stats);
+                        future::ok::<_, futures_mpsc::SendError<_>>((resp, (ctx, finished)))
+                    })
+                });
+                on_resp.respond_stream(box s, pool.clone());
+            }
+            CopRequest::Analyze(analyze) => {
+                let ctx = AnalyzeContext::new(analyze, ranges, snap, &req_ctx);
+                let mut cop_stats = CopStats::default();
+                let do_request = move || {
+                    tracker.record_wait();
+                    let mut resp = ctx.handle_request(&mut cop_stats.stats)
+                        .unwrap_or_else(err_resp);
+                    tracker.record_handle(&mut resp, cop_stats);
+                    future::ok::<_, ()>(on_resp.respond(resp))
+                };
+                pool.spawn_fn(do_request).forget();
+            }
+        }
     }
 
     fn handle_snapshot_result(&mut self, id: u64, snapshot: engine::Result<Box<Snapshot>>) {
-        let reqs = self.reqs.remove(&id).unwrap();
         let snap = match snapshot {
             Ok(s) => s,
-            Err(e) => {
-                notify_batch_failed(e, reqs);
-                return;
-            }
+            Err(e) => return self.notify_batch_failed(e, id),
```

Comment:
So if there is an error, `reqs` never gets cleaned up?

Reply:
I think we do the clean up in `notify_batch_failed`?